website/src/components/Stats/Stats.components.tsx

```diff
@@ -38,30 +39,208 @@ type MessageTreeStateStatsProps = {
+    "growing",
+    "backlog_ranking",
+    "ranking",
+    "ready_for_export",
+    "aborted_low_grade",
+    "halted_by_moderator",
+  ];
+
+  const languages = [];
+  for (let i = 0; i < stat.stats.length; i++) {
+    if (!languages.includes(stat.stats[i].lang)) {
+      languages.push(stat.stats[i].lang);
+    }
+  }
+
+  const rows = [];
+
+  for (let i = 0; i < stat.stats.length; i++) {
+    const lang = stat.stats[i].lang;
+
+    let langRow = findLangRow(rows, lang);
+
+    if (langRow === -1) {
+      // The language is not yet in the data. Add it
+      rows.push(Array(headers.length).fill(0));
+      langRow = rows.length - 1;
+      rows[langRow][0] = lang;
+    }
+
+    const state = stat.stats[i].state;
+    const stateCol = headers.indexOf(state);
+    if (stateCol === -1) {
+      continue;
+    }
+
+    const threadCount = stat.stats[i].count;
+    rows[langRow][stateCol] = threadCount;
+  }
+
+  headers = headers.map((item) => t(getTypeSafei18nKey(item)));
+
+  for (let i = 0; i < rows.length; i++) {
+    const langCode = rows[i][0];
+    rows[i][0] = getLocaleDisplayName(langCode) + " [" + langCode + "]";
+  }
+  return { headers, rows };
+}
+
 export const MessageTreeStateStats = ({ stat, titleFn }: MessageTreeStateStatsProps) => {
   const { t } = useTranslation("stats");
+  const color = useColorModeValue(colors.light.text, colors.dark.text);
+  const langs = stat.stats.map((item) => item.lang);
+  const uniqueLangs = langs.filter((item, index) => langs.indexOf(item) === index);
+  const [selectedLang, setSelectedLang] = useState("en");
+  const stats = stat.stats.filter((item) => item.lang === selectedLang);
+
+  return (
+    <>
+      <Select cursor="pointer" mb={3} value={selectedLang} onChange={(e) => setSelectedLang(e.target.value)}>
+        {uniqueLangs.map((lang) => (
+          <option key={lang} value={lang}>
+            {titleFn(lang)}
+          </option>
+        ))}
+      </Select>
+      <Box minH={330}>
+        <Doughnut
+          width={100}
+          height={50}
+          options={getDefaultChartOptions(color)}
+          data={{
+            labels: stats.map((item) => t(getTypeSafei18nKey(item.state))),
+            datasets: [
+              {
+                data: stats.map((item) => item.count),
+              },
+            ],
+          }}
+        />
+      </Box>
+    </>
+  );
+};
+
+export const MessageTreeStateStatsStacked = ({ stat, titleFn }: MessageTreeStateStatsProps) => {
+  const { t } = useTranslation("stats");
+  const langs = stat.stats.map((item) => item.lang);
+  const uniqueLangs = langs.filter((item, index) => langs.indexOf(item) === index);
+  const states = stat.stats.map((item) => item.state);
+  const uniqueStates = states.filter((item, index) => states.indexOf(item) === index);
+
+  const bar_options = {
+    responsive: true,
+    scales: {
+      x: {
+        stacked: true,
+      },
+      y: {
+        stacked: true,
+      },
+    },
+  };
```

Comment:
This is constant so it could be moved out.

Reply:
Sorry, I don't understand what you are proposing. Moving it out of where and to where?

Reply:
@GuilleHoardings He mean move the bar_options to top-level scope (out side of MessageTreeStateStatsStacked component)

Reply:
Thanks, I see. However, as it only used in that function, isn't it better to leave inside? I see it as a non-mutable variable, which is defined where is used, not as a constant for all the program.

Reply:
The main reason I would move this out is so that it is not declared on every render of the component.

Reply:
That makes sense. I was not aware of that kind of performance problems. Thank you!